drivers/rtc/rtc_stm32.c

```diff
@@ -0,0 +1,181 @@
+	timeDef.Seconds = tm.tm_sec;
+	if (LL_RTC_TIME_Init(RTC, LL_RTC_FORMAT_BIN, &timeDef) != SUCCESS) {
+		serr = "failed to set time";
+		goto out;
+	}
+
+	dateDef.WeekDay = ((tm.tm_wday + 6) % 7) + 1;
+	dateDef.Month = tm.tm_mon + 1;
+	dateDef.Day = tm.tm_mday;
+	dateDef.Year = (tm.tm_year + 1900) - 2000;
+	if (LL_RTC_DATE_Init(RTC, LL_RTC_FORMAT_BIN, &dateDef) != SUCCESS) {
+		serr = "failed to set date";
+		goto out;
+	}
+
+	err = 0;
+out:
+	LL_PWR_DisableBkUpAccess();
+
+	irq_unlock(key);
+
+	if (serr) {
+		LOG_ERR("%s", serr);
+	}
+
+	return err;
+}
+
+static int rtc_stm32_get_time(struct device *dev, struct timespec *tp)
+{
+	struct tm tm;
+	u32_t time;
+	u32_t subSeconds;
+	u32_t date;
+	int key;
+
+	ARG_UNUSED(dev);
+
+	key = irq_lock();
+	subSeconds = LL_RTC_TIME_GetSubSecond(RTC);
+	time = LL_RTC_TIME_Get(RTC);
+	date = LL_RTC_DATE_Get(RTC);
+	irq_unlock(key);
+
+	tm.tm_sec = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_SECOND(time));
+	tm.tm_min = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_MINUTE(time));
+	tm.tm_hour = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_HOUR(time));
+
+	tm.tm_mday = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_DAY(date));
+	tm.tm_mon = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_MONTH(date)) - 1;
+	tm.tm_year = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_YEAR(date)) + 100;
+	tp->tv_sec = timeutil_timegm(&tm);
+	tp->tv_nsec = ((u64_t)(RTC_PREDIV_SYNC - subSeconds) * NSEC_PER_SEC) /
+		(RTC_PREDIV_SYNC + 1);
+
+	return 0;
+}
+
+static int rtc_stm32_init(struct device *dev)
+{
+	int err = -EIO;
+	struct device *clk;
+	struct stm32_pclken pclken;
+	LL_RTC_InitTypeDef init;
+
+	clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	__ASSERT_NO_MSG(clk);
+
+	pclken.enr = DT_INST_0_ST_STM32_RTC_CLOCK_BITS;
+	pclken.bus = DT_INST_0_ST_STM32_RTC_CLOCK_BUS;
+	clock_control_on(clk, (clock_control_subsys_t *) &pclken);
+
+	LL_PWR_EnableBkUpAccess();
+
+#if defined(CONFIG_RTC_STM32_CLOCK_LSI)
+	LL_RCC_LSI_Enable();
+
+	/* Wait until LSI is ready */
+	while (LL_RCC_LSI_IsReady() != 1) {
+	}
+
+	LL_RCC_SetRTCClockSource(LL_RCC_RTC_CLKSOURCE_LSI);
+
+#else /* CONFIG_RTC_STM32_CLOCK_LSE */
+	LL_RCC_LSE_Enable();
+
+	/* Wait until LSE is ready */
+	while (LL_RCC_LSE_IsReady() != 1) {
+	}
+
+	LL_RCC_SetRTCClockSource(LL_RCC_RTC_CLKSOURCE_LSE);
+
+#endif
+
+	LL_RCC_EnableRTC();
+
+	init.HourFormat = RTC_HOURFORMAT_24;
+	init.AsynchPrescaler = RTC_PREDIV_ASYNC;
+	init.SynchPrescaler = RTC_PREDIV_SYNC;
+	if (LL_RTC_Init(RTC, &init) != SUCCESS) {
+		LOG_ERR("failed to init");
+		goto out;
+	}
+	/* check if calendar has been initialized */
+	if (!LL_RTC_IsActiveFlag_INITS(RTC)) {
+		LOG_INF("Datetime initialization");
+		struct timespec tp;
+		/* Init date time to 1 January 2001 00:00:00 */
+		tp.tv_sec = 978307200;
+		tp.tv_nsec = 0;
+		err = rtc_stm32_set_time(dev, &tp);
+		if (err < 0) {
+			goto out;
+		}
+		/* BkUpAccess is disabled by rtc_stm32_set_time */
+		LL_PWR_EnableBkUpAccess();
+	}
+
+	err = 0;
+out:
+	LL_PWR_DisableBkUpAccess();
```

Comment:
Looks a little bit strange to enable something just to be able to disable it. Wouldn´ t an else path be more appropriate?